keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1297 @@
+                        obj, err := legacyscheme.Scheme.New(gv.WithKind(kind))
+                        if err != nil {
+                                t.Fatalf("could not create a %v: %s", kind, err)
+                        }
+                        f.Fuzz(obj)
+                        verifyVersionedValidationEquivalence(t, obj, nil)
+
+                        old, err := legacyscheme.Scheme.New(gv.WithKind(kind))
+                        if err != nil {
+                                t.Fatalf("could not create a %v: %s", kind, err)
+                        }
+                        f.Fuzz(old)
+                        verifyVersionedValidationEquivalence(t, obj, old)
+                }
+        }
+}
+```
+
+###### strategy_test.go vs validation_test.go
+<<[UNRESOLVED should we move current validation_test.go logic to strategy_test.go as it makes more sense generally and aids this project? ]>>
+
+
+Currently, validation logic and associated tests are logically split across `validation.go` and `strategy.go`. The hand-written validation functions and associated tests reside in `validation.go` and `validation_test.go`  while `strategy.go` determines when to invoke these functions during API object creation and updates. With the introduction of declarative validation (controlled by the `DeclarativeValidation` feature gate)  the current logic split is worth considering moving from `validation_test.go` to `strategy_test.go` as the current logic split in the test is unfavorable for the migration as it requires additional plumbing work.
+
+The current approach in the prototype experiment to migrate ReplicationController involves directly injecting declarative validation and equivalency tests into `validation_test.go`. This is achieved by conditionally appending calls to `rest.ValidateDeclaratively` within the existing test cases, based on the `DeclarativeValidation` feature gate's status. This allows for a direct comparison of the outputs between hand-written and declarative validation within the same test framework.
+
+**Moving the feature gate check and declarative validation logic to `strategy_test.go` could offer several advantages</strong>:
+
+*   **Reduced Test Duplication:** `validation_test.go` could be simplified, as it would no longer need to handle both hand-written and declarative validation paths.
+*   **Clearer Separation of Concerns:** `strategy.go` would be responsible for determining _when_ to validate, while `validation.go` would handle _how_ to validate.
+*   **Easier Migration:**  Transitioning to a fully declarative model would be smoother, as the core validation logic would already be invoked through the strategy.
+
+However, this approach also has potential drawbacks:
+
+*   **Additional PRs for Initial Migration:** Moving existing hand-written validation logic from `validation.go` to `strategy.go` would result in additional work via a PR to move the related code but it would be straightforward (only moving files).
```

Comment:
it also moves unit testing from the package that is being unit tested to a different package. In addition to being unusual, it means code coverage calculations are wrong.

I don't mind the switch if this is what @jpbetz and @thockin prefer, but I do want it to be resolved in the KEP prior to merge.

Reply:
I've updated this section to explain our current plan and explain that we will switch where the tests are currently located (validation_test.go -> strategy_test.go).  In a previous discussion w/ @jpbetz and @thockin the preferred approach to doing this would be immediately after the initial v1.33 PR before any other migrations are done.  I've updated the text here to state this and removed the UNRESOLVED